### DISCUSSION TITLE
Is it feasible to use prisma client in the browser against sqlite compiled to webassembly

### DISCUSSION THREAD

**Discussion by chuckdries** [prisma/prisma#18428] [Q&A] [answered]
Hello. My project uses Prisma with sqlite on the backend already, and we're quite happy with it. I'm now working on implementing offline support in our client, and it occurred to me that it'd be great to be able to reuse some of our schema between the client and the server (or at least get to use some of prisma's awesome features in both places if schema can't be shared). I'm looking for a sanity check that this is even possible today, and if it is, guidance on where to start. I see that the official sqlite project has support for running in the browser with webassembly and storage provided by localstorage. We've already got a monorepo set up that allows us to share JS code successfully between the client and server. 

If this idea isn't possible today, perhaps this post can be a place of discussion for what it might take to make it possible in the future, and whether the project is interested in doing so.

**Comment by janpio** [comment, marked as answer]
Prisma currently can not do that, as we do not compile the full engines to Wasm.
Attempting this is also not a priority right now, as we are quite busy with Node as a environment - we won't add browser to our workload here at least this year.

**Reply by lastmjs** [reply]
Is this still the case? Any update? Where can I get more information on [@prisma/query-engine-wasm](https://www.npmjs.com/package/@prisma/query-engine-wasm)?

**Reply by janpio** [reply]
No update, the package you linked to is used for https://www.prisma.io/blog/prisma-orm-support-for-edge-functions-is-now-in-preview. It is used in the code of this repository, so if you want to see if you can also adapt this for a browser you can start digging here in the TS codebase.

**Reply by AntonOfTheWoods** [reply]
Electric-sql uses a modified Prisma 4.8.1 generated client to talk to to both wa-sql and pglite in the browser. It might be possible to work with them to rip out and generify the relevant parts.

**Reply by juemrami** [reply]
Is this still not a priority for prisma?
Id love to have prisma as the contract between the browsers local db and server's copy of the db. Would be phenomenal for local-first support. Would also be great to have prisma's migration system as a way to manage users offline/local db version upgrading. 

There's also some interest brewing in this issue:
- #5385

**Comment by clicktodev** [comment]
considering prisma 7 client is rust free, i believe it should be possible to use prisma in the browser now

**Reply by zwhitchcox** [reply]
Is anyone working on this? I would be interested in trying it out/contributing

**Comment by WhyAsh5114** [comment]
I ran into this same problem and ended up building a generator for it — [prisma-idb](https://prisma-idb.dev/). It uses IndexedDB as the storage layer instead of SQLite/WASM, but gives you the same Prisma Client API in the browser, generated from your existing schema. Supports findMany, include, nested writes, etc.

If you specifically need SQLite semantics or raw SQL, it won't fit. But if the goal is Prisma API + offline storage + optional bidirectional sync, it might be worth a look.

**Comment by rich-row-chat** [comment]
Prisma has a WASM query builder and now makes this practical, so we published: @row-chat/prisma-sqlite-wasm-adapter, an in-browser Prisma adapter for @sqlite.org/sqlite-wasm.

- Source: https://github.com/row-chat/prisma-browser-adapters                                                                                   
- Live demo: https://row-chat.github.io/prisma-browser-adapters/ — the Chinook sample database running entirely in the browser, exposed through a Prisma Client REPL, a raw SQL REPL, and an embedded Prisma Studio. SQLite-WASM persists the database in OPFS via the SAH pool VFS.
                                 
How it works: the sqlite-wasm DB lives in a SharedWorker (so all tabs on the same origin see the same database). Queries like prisma.album.findMany() work entirely in the browser.

**Comment by rich-row-chat** [comment]
We also have working in-browser adapters for wa-sqlite and PGLite that we plan to publish in the future.

**Comment by rich-row-chat** [comment]
The Prisma adapter for wa-sqlite has been published: https://www.npmjs.com/package/@row-chat/prisma-wa-sqlite-adapter

**Comment by rich-row-chat** [comment]
PGLite can be used with Prisma in the browser without extra dependencies. Here are the instructions. With the plain PGLite Prisma adapter each browser tab uses a full instance of PGLite. In the future the package described in this link will be published to enable one instance of PGLite to coordinate through a SharedWorker: https://github.com/row-chat/prisma-browser-adapters/tree/main/packages/prisma-pglite-adapter#readme
